### PR TITLE
[DSM] DDP-8802 | Tweaking "customize view columns" for old Osteo study

### PR DESCRIPTION
**Pull request by ScalaChamp**
#### **Things have done**:

 - Establish a new `enum` called `OldOsteoActivityCode.java` for grouping the set of activity codes for old `Osteo` study.
 - Add a new generic predicate based interface called `StudyPostFilterStrategy.java` as a marker for specific implementations.
 - Create `NewOsteoPostFilterStrategy.java` and `OldOsteoPostFilterStrategy.java` implementations for new and old `Osteo` instances respectively.
 - Model a new generic interface called `ElasticSearchObjectsRetriever<A>` which can be the base high level description of retrieving any kind of data from ElasticSearch.
 - Set up two implementations for the aforementioned interface:
   - `ActivityDefinitionsRetriever.java` which will serve es a default implementation for each study.
   - `OldOsteoActivityDefinitionsRetriever.java` which will serve as an implementation for old `Osteo` study.
 - Design a new factory for instances of `ActivityDefinitionsRetriever.java` which can return one of its two sub-classes based on the name of the `DdpInstance`.
 - Modify the `DisplaySettingsRoute.java` with the help of new services.
 - Write `unit` and `property-based` tests for new services, modify old unit tests.
  
_Say a few words about the bigger picture beyond this PR. How do your changes improve life for our users? Link directly to the jira ticket (just type "DDP-..." and it'll automatically link) to help your teammates quickly get up to speed on the motivation for these changes._

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/retrieve/activities/ActivityDefinitionsRetriever.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    private final DDPInstance instance;
+
+    public ActivityDefinitionsRetriever(DDPInstance instance) {
+        this.instance = instance;
+    }
+
+    @Override
+    public Map<String, Map<String, Object>> retrieve() {
+        Map<String, Map<String, Object>> activityDefinitions =
+                new ConcurrentHashMap<>(ElasticSearchUtil.getActivityDefinitions(instance));
```
Comment (gmakharat) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/retrieve/activities/ActivityDefinitionsRetriever.java]: I think it is good idea to move `getActivityDefinitions` method to `ElasticSearchable` and `ElasticSearch` to eliminate `ElasticSearchUtil` and `DDPInstance` usage.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/filter/postfilter/osteo/OldOsteoActivityCode.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    PARENTAL_CONSENT,
+    RELEASE_MINOR,
+    RELEASE_SELF,
+    PREQUAL,
+    ABOUT_YOU,
+    CONSENT,
+    LOVEDONE,
+    CONSENT_ASSENT;
+
+    public static String[] asStrings() {
```
Comment (gmakharat) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/filter/postfilter/osteo/OldOsteoActivityCode.java]: I think returning array of Strings kinda breaks the idea of enums, as I see these values are used in steam for boolean value, I would create method something like `isOldOsteoActivityCode` and I would use `valueOf` method in it wrapped by `try/catch` and would return boolean value.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/filter/postfilter/osteo/OldOsteoPostFilterStrategy.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.broadinstitute.dsm.model.filter.postfilter.StudyPostFilterStrategy;
+
+public class OldOsteoPostFilterStrategy implements StudyPostFilterStrategy<Activities> {
+
+    private static final String[] OLD_OSTEO_ACTIVITY_CODES = OldOsteoActivityCode.asStrings();
+    private static final String OLD_OSTEO_ACTIVITY_VERSION = "v1";
+
+    @Override
+    public boolean test(Activities activities) {
+        return Arrays.stream(OLD_OSTEO_ACTIVITY_CODES)
```
Comment (gmakharat) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/filter/postfilter/osteo/OldOsteoPostFilterStrategy.java]: here you would have `return OldOsteoActivityCode .isOldOsteoActivityCode(activities.getActivityCode()) && activities.getActivityVersion().equals(OLD_OSTEO_ACTIVITY_VERSION)`

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/retrieve/activities/ActivityDefinitionsRetriever.java (hunk 1 above)
Comment (ScalaChamp) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/retrieve/activities/ActivityDefinitionsRetriever.java]: @gmakharat I agree with defining this method in interface and making it as a `default` implementation / implementing in `ElasticSearch`. 

However there's one problem with removing `DDPInstance`. If you look at `DisplaySettingsRoute.java` you will see that `DDPInstance` is used in many places.

So I belive I have two options:
  - use `DDPInstance` as a dependency for the new method in `ElasticSearchable`
  - use `DDPInstanceDto` as a dependecy for the new method in `ElasticSearchable` and refactor the whole `DisplaySettingsRoute.java`
 
I would like to know your point of view on this, thxs!

Comment (gmakharat) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/retrieve/activities/ActivityDefinitionsRetriever.java]: depends how widely `DDPInstance` is used in `DisplaySettingsRoute`, if it is easy change, theნ it worths to change `DDPInstance` to `DDPInstanceDto`, if not, let's keep it how it is right now.

**Review by gmakharat — DISMISSED** [review]

**Review by GiCharkviani — APPROVED** [review]

**Pull request merged**